app/src/main/java/org/simple/clinic/teleconsultlog/teleconsultrecord/screen/TeleconsultRecordEffectHandler.kt

```diff
@@ -52,20 +52,23 @@ class TeleconsultRecordEffectHandler @AssistedInject constructor(
       effects
           .observeOn(schedulersProvider.io())
           .map { effect -> prescriptionRepository.newestPrescriptionsForPatientImmediate(effect.patientUuid) to effect }
-          .filter { (prescriptions, _) -> prescriptions.isNotEmpty() }
-          .doOnNext { (prescriptions, _) -> prescriptionRepository.softDeletePrescriptions(prescriptions) }
           .flatMap { (prescriptions, effect) ->
-            val clonedPrescriptions = prescriptions.map { prescribedDrug ->
-                  prescribedDrug.refillForTeleconsultation(
-                      uuid = uuidGenerator.v4(),
-                      facilityUuid = currentFacility.get().uuid,
-                      teleconsultationId = effect.teleconsultRecordId,
-                      utcClock = utcClock
-                  )
-                }
+            if (prescriptions.isNotEmpty()) {
+              prescriptionRepository.softDeletePrescriptions(prescriptions)
 
-            prescriptionRepository.save(clonedPrescriptions)
-                .andThen(Observable.just(PatientPrescriptionsCloned))
+              val clonedPrescriptions = prescriptions.map { prescribedDrug ->
+                prescribedDrug.refillForTeleconsultation(
+                    uuid = uuidGenerator.v4(),
+                    facilityUuid = currentFacility.get().uuid,
+                    teleconsultationId = effect.teleconsultRecordId,
+                    utcClock = utcClock
+                )
+              }
+
+              prescriptionRepository.saveImmediate(clonedPrescriptions)
+            }
```

Comment:
This can be moved to a `.doOnNext {}` block, right?

Reply:
The Rx chain then becomes a little simpler since we can avoid the `.flatMap {}`.

```kotlin
.doOnNext { /* Delete old prescriptions, clone and save new ones */ }
.map { PatientPrescriptionsCloned } 
```

Reply:
Don't we have to wait until cloning is completed before moving to next screen? 🤔

Reply:
`.doOnNext` will wait.

Reply:
Operators wait unless you explicitly change the scheduler using a `.flatMap {}`.

Reply:
Ah okay, I think I got confused with the terminology. I thought being side effect meant it will not wait for the operations to be completed, but actually it will wait but doesn't send the results to downstream right?

Reply:
Yup.